pyrenew/time.py

```diff
@@ -0,0 +1,258 @@
+       the informativeness of the error message.
+
+    Returns
+    -------
+    None
+       If validation passes.
+
+    Raises
+    ------
+    ValueError
+       If validation fails.
+    """
+    if not isinstance(day_of_week, int):
+        raise ValueError(
+            "Day-of-week indices must be integers "
+            "between 0 and 6, inclusive. "
+            f"Got {day_of_week} for {variable_name}, "
+            "which is a "
+            f"{type(day_of_week)}"
+        )
+    if day_of_week < 0 or day_of_week > 6:
+        raise ValueError(
+            "Day-of-week indices must be a integers "
+            "between 0 and 6, inclusive. "
+            f"Got {day_of_week} for {variable_name}."
+        )
+    return None
+
+
+def daily_to_weekly(
+    daily_values: ArrayLike,
+    input_data_first_dow: int = 0,
+    week_start_dow: int = 0,
+) -> ArrayLike:
+    """
+    Aggregate daily values (e.g.
+    incident hospital admissions)
+    to weekly total values.
+
+    Parameters
+    ----------
+    daily_values : ArrayLike
+        Daily timeseries values (e.g. incident infections or
+        incident ed visits).
+    input_data_first_dow : int
+        First day of the week in the input timeseries `daily_values`.
+        An integer between 0 and 6, inclusive (0 for Monday, 1 for Tuesday,
+        ..., 6 for Sunday).
+        If `input_data_first_dow` does not match `week_start_dow`, the
+        incomplete first week is ignored and weekly values starting
+        from the second week are returned. Defaults to 0.
+    week_start_dow : int
+        Day of the week on which weeks are considered to
+        start in the output timeseries of weekly values
+        (e.g. ISO weeks start on Mondays and end on Sundays;
+        MMWR epiweeks start on Sundays and end on Saturdays).
+        An integer between 0 and 6, inclusive (0 for Monday,
+        1 for Tuesday, ..., 6 for Sunday).
+        Default 0 (i.e. ISO weeks, starting on Mondays).
+
+    Returns
+    -------
+    ArrayLike
+        Data converted to weekly values starting
+        with the first full week available.
+
+    Raises
+    ------
+    ValueError
+        If the specified days of the week fail validation.
+
+    Notes
+    -----
+    This is _not_ a simple inverse of :func:`weekly_to_daily`.
+    This function aggregates (by summing) daily values to
+    create a timeseries of weekly total values.
+    :func:`weekly_to_daily` broadcasts a _single shared value_
+    for a given week as the (repeated) daily value for each day
+    of that week.
+    """
+
+    validate_dow(input_data_first_dow, "input_data_first_dow")
+    validate_dow(week_start_dow, "week_start_dow")
+
+    offset = (week_start_dow - input_data_first_dow) % 7
+    daily_values = daily_values[offset:]
+
+    if len(daily_values) < 7:
+        raise ValueError("No complete weekly values available")
+
+    weekly_values = jnp.convolve(daily_values, jnp.ones(7), mode="valid")[::7]
```

Comment:
```suggestion
    trimmed = daily_values[:len(daily_values) // 7 * 7]
    weekly_values = trimmed.reshape(-1, 7).sum(axis=1)
```
Seems to be about 5x faster

Reply:
Nice, thanks! As written this function only accepts 1D inputs and will error with a greater than 1D input (because `jnp.convolve` does). The new version also assumes 1D but doesn't verify. We can verify 1D input but even better would be to generalize.

Reply:
```suggestion
    n_weeks = daily_values.shape[0] // 7
    trimmed = daily_values[:n_weeks * 7] 
    return trimmed.reshape(n_weeks, 7, *daily_values.shape[1:]).sum(axis=1)
```

This works, assuming the 0th dimension is "time."

Reply:
Thanks! Would you be up for contributing an edit to the tests to check the multi-dim (1st dim time) case?

Reply:
@dylanhmorris pushed directly to this branch [`7a85518` (#483)](https://github.com/CDCgov/PyRenew/pull/483/commits/7a85518e143c77e584306fa46274fb579b6ed8ed)

Reply:
thanks!